### PR TITLE
(#204) applicant changes

### PR DESCRIPTION
**Pull request by rathorevaibhav**
#204 

Changes include:

1. Migration for deleting old columns `hr_applicant_id` from tables. After changes implemented in #210. These columns were needed to migrate data only and are no longer required.
2. Check for an applicant if it already exists.
3. Modified the workflow around applicant and application generation.
4. Modified the workflow around application round actions.
5. In case an applicant applies for multiple jobs, the first job is marked as in-progress. All the following will be marked as on-hold.
6. In case an applicant applies for multiple jobs, the reject action will open up a modal. That modal will have options either to reject for this application but refer for another appliction or to reject for all applications at once.

### CODE REVIEW

File: app/Http/Controllers/HR/ApplicationController.php [hunk 1]
```diff
@@ -109,4 +109,24 @@ public function destroy(Application $application)
     {
         //
     }
+
+    /**
+     * get list of applications based on their show status
+     * @return \Illuminate\Database\Eloquent\Collection
+     */
+    public function getApplicationsList()
```
Comment (bohraji) [inline app/Http/Controllers/HR/ApplicationController.php]: This piece is pretty specific to the `Application` model. Can we move it to the model and have it called here. Any specific changes that is required to be done to the data that has been retrieved, before it is passed to the controller method to work with views, can be done in this method.

Also, there is a possibility that we'll be by default fetching the non reject application, so we can have it as default.

File: app/Http/Controllers/HR/ApplicationController.php [hunk 2]
```diff
@@ -109,4 +109,24 @@ public function destroy(Application $application)
...
+    {
+        $request = request();
+        $applications = Application::with('applicant', 'job');
+        $rejected = config('constants.hr.status.rejected.label');
+        switch ($request->get('status')) {
+            case $rejected:
+                $applications = $applications->where('status', $rejected);
```
Comment (bohraji) [inline app/Http/Controllers/HR/ApplicationController.php]: I see the possibility of using this filter a lot. We have a local scope for it. That would make it much more readable. Thoughts?

File: app/Http/Controllers/HR/ApplicationController.php [hunk 3]
```diff
@@ -109,4 +109,24 @@ public function destroy(Application $application)
...
+                break;
+            default:
+                $applications = $applications->where('status', '!=', $rejected);
+                break;
+        }
+        return $applications->orderBy('id', 'desc')->paginate(config('constants.pagination_size'));
```
Comment (bohraji) [inline app/Http/Controllers/HR/ApplicationController.php]: Also, the list that we retrieve is mostly paginated and ordered. We can add that as a global or local scope. Thoughts?

File: app/Models/HR/Applicant.php [hunk 4]
```diff
@@ -23,39 +20,35 @@ class Applicant extends Model
...
-            'reviews' => $request->input('reviews'),
-        ]));
-        return $updated;
-    }
-
-    public function getApplicantRound($round_id)
+    public function openApplications()
     {
-        return $this->applicantRounds->where('hr_round_id', $round_id)->first();
+        return $this->applications->where('status', '!=', config('constants.hr.status.rejected.label'));
```
Comment (bohraji) [inline app/Models/HR/Applicant.php]: We can use the model scope defined above here.

**Comment by bohraji** [comment]
Checked this PR :

- [ ] For duplication.
- [ ] For performance. 
- [ ] Usage of recursive methods, loops and algorithms.
- [ ] For readability and understanding. 
- [ ] For functions with multiple responsibilities.
- [ ] For scope of refactoring. 
- [ ] For coupling
- [ ] For alternative approaches to the implementation. 
- [ ] For how the change impacts other parts of the system.
- [ ] For validations. client side and server side validations are present.

File: app/Http/Controllers/HR/ApplicationController.php (hunk 3 above)
Comment (rathorevaibhav) [reply app/Http/Controllers/HR/ApplicationController.php]: Agreed. It makes sense. Will proceed with the changes.

Comment (rathorevaibhav) [reply app/Http/Controllers/HR/ApplicationController.php]: I think a global scope will probably make more sense here. Considering that, the changes will happen in other modules as well. I think we should target it separately. @bohraji thoughts?

**Pull request merged**